BUG-22984 Fix truncation of DataFrame representations

JustinZhengBC opened the pull request:
- [X] closes #22984 
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

When printing a DataFrame to terminal, an extra column's worth of space is added to the calculated width of the DataFrame. This is presumably to help edge cases, but the calculated difference between the DataFrame width and the terminal window width is incremented by 1 a few lines later, seemingly to fix the same problem. Do any more experienced developers know of a reason to pad the DataFrame width even more?

JustinZhengBC committed:
BUG-22984 Fix truncation of DataFrame representations

JustinZhengBC committed:
BUG-22984 Fix flake8 issues

pep8speaks commented:
Hello @JustinZhengBC! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/io/formats/format.py`](https://github.com/JustinZhengBC/pandas/blob/9526ec7a9667b259849e13979b01117370b53ddc/pandas/io/formats/format.py) !

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -194,6 +194,7 @@ Other Enhancements
 - :meth:`Index.to_frame` now supports overriding column name(s) (:issue:`22580`).
 - New attribute :attr:`__git_version__` will return git commit sha of current build (:issue:`21295`).
 - Compatibility with Matplotlib 3.0 (:issue:`22790`).
+- Representation of :class:`DataFrame` fills up the terminal window better
```
can you add the issue number here, I would call this more of a bug fix, no?

jreback commented on pandas/io/formats/format.py:
```diff
@@ -616,11 +616,6 @@ def to_string(self):
             else:  # max_cols == 0. Try to fit frame to terminal
                 text = self.adj.adjoin(1, *strcols).split('\n')
                 max_len = Series(text).str.len().max()
-                headers = [ele[0] for ele in strcols]
```
we have a number of tests for this, suprised this didn't break anything, can you make a test if we have no coverage for this case now? (e.g. you use option_context to set the width, then check the output string)

jreback commented:
cc @jorisvandenbossche

JustinZhengBC committed:
BUG-22984 Fix whatsnew and add test

JustinZhengBC committed:
BUG-22984 Fix whatsnew and add test

JustinZhengBC committed:
BUG-22984 Fix linting issue

HubertKl committed:
DOC: Updating the docstring of Series.dot  (#22890)

datapythonista committed:
Fixing flake8 problems new to flake8 3.6.0 (#23472)

jbrockmendel committed:
strictness and checks for Timedelta _simple_new (#23433)

jbrockmendel committed:
REF: cython cleanup, typing, optimizations (#23464)

* Easy bits of #23382

* Easy parts of #23368

gfyoung committed:
ENH: Add FrozenList.union and .difference (#23394)

Re-attempt of gh-15506.

Closes gh-15475.

TomAugspurger committed:
BUG: Allow freq conversion from dt64 to period (#23460)

Closes #23438

hongshaoyang committed:
add number of Errors, Warnings to scripts/validate_docstrings.py (#23150)

Commit:
DOC: Add cookbook entry for triangular correlation matrix (GH22840) (#23032)

topper-123 committed:
CLN: doc string (#23469)

jbrockmendel committed:
STYLE: Standardize cython spacing for casting, with linting (#23474)

Commit:
DOC: Adding documentation for pandas.core.indexes.api internal functions (#22980)

bartaelterman committed:
DOC: Clarify documentation of 'ambiguous' parameter (#23408)

* Add documentation line with example for the ambiguous parameter of tz_locaclize

* Updating 'ambiguous'-param doc + update it on Timestamp, DatetimeIndex and NaT

This is following the discussion at
https://github.com/pandas-dev/pandas/pull/23408#discussion_r229109984

thoo committed:
DOC: Validate in docstrings that numpy and pandas are not imported (#23161)

eldritchideen committed:
DOC: Updated docstrings related to DateTimeIndex. GH22459 (#22504)

dragosthealex committed:
DOC: Rephrased doc for Series.asof. Added examples (#21034)

mroeschke committed:
TST: drop tz-aware timestamp from DatetimIndex with DST transition (#23479)

ZackStone committed:
DOC: update the pandas.Series.shift docstring (#20472)

alimcmaster1 committed:
Isort contributing guide (#23364)

h-vetinari committed:
API: fix corner case of lib.infer_dtype (#23422)

FHaase committed:
DOC: Use flake8 to check for PEP8 violations in doctests (#23399)

mroeschke committed:
TST: Empty Series.reindex with tz-aware dtype (#23480)

vkk800 committed:
DOC: Fix creation of [source] links in the doc creation (#23129)

mroeschke committed:
DOC: Remove dead link and update links to https (#23476)

* DOC: Remove dead links and update links to https

* Add missing ~

Moisan committed:
DOC: Fix DataFrame.nlargest and DataFrame.nsmallest doctests (#23202)

gfyoung committed:
DOC: Fix syntax error in groupby docs (#23498)

Follow-up to gh-23394.

thoo committed:
Run Isort on tests-> util,sereis,arrays (#23501)

gfyoung committed:
TST: Add test of assignment chaining and dupe cols (#23487)

xref gh-13017.

icexelloss committed:
BUG: Avoid casting to double type unnecessarily when setting values i… (#23462)

anjsudh committed:
Bumping up min version for pyarrow and fastparquet (#23482)

* Bumping up min version for pyarrow

TomAugspurger committed:
CI: Fixed pytest minversion (#23520)

qwhelan committed:
CI: Exclude asv benchmark envs from flake8 checks (#23497)

jbrockmendel committed:
Extraneous parts broken off from other PRS (#23518)

datapythonista committed:
Fixing typo in cython casting lint, and making it azure friendly (#23486)

TomAugspurger committed:
CI: Unpin NumPy (#23465)

gfyoung committed:
TST: Add test for mangling of unnamed columns (#23485)

xref gh-13017.

jbrockmendel committed:
PERF: cython optimizations (#23477)

sinhrks committed:
BUG: Cleanup timedelta offset (#23439)

h-vetinari committed:
API: change default for sep in str.cat (in docstring) (#23443)

pajachiet committed:
BUG: fix df.where(cond) when cond is empty (#21947)

gfyoung committed:
BUG: Fix of handle missing CSV MI column names (#23484)

JustinZhengBC committed:
BUG GH23451 Allow setting date in string index for Series (#23495)

DiegoAlbertoTorres committed:
BUG: fix groupby.transform rename bug (#23461) (#23463)

5hirish committed:
BUG: GroupBy return EA dtype (#23318)

mitstake committed:
BUG - pd.concat with all Series on axis=1 ignores the `names` argument (#23499)

Closes gh-23490.

Dr-Irv committed:
BUG: names on union and intersection for Index were inconsistent (#19849)

Closes gh-9862.

xref gh-9943.

TomAugspurger committed:
DOC/TST: Fix warning in concat docstring (#23529)

xref https://github.com/pandas-dev/pandas/issues/23149

datapythonista committed:
CI: Moving CI dependencies to "ci/deps" (#23526)

jbrockmendel committed:
Fix+test timedelta64(nat) ops (#23425)

Follow-up to gh-23320.

alimcmaster1 committed:
CLN: Isort io and compat dirs (#23534)

alimcmaster1 committed:
CI: Fail Azure CI for failing tests (#23454)

datapythonista committed:
Make validate_docstrings.py ready for the CI (#23514)

* validate_docstrings.py to exit with status code as the number of errors (so, 0 for no errors)

* Implemented different output types for the validate_all, and a prefix to filter which docstrings are validated

* Codifying errors

* Adding --errors parameter to be able to validate only specific errors

datapythonista committed:
Fixing bug in validate_docstrings.py, where a bracket closed in the wrong place, raised a KeyError when a private class was found in a docstrings (and adding test for that case (#23543)

mroeschke committed:
BUG: Timestamp retains frequency of input Timestamps (#23503)

sinhrks committed:
BUG: DatetimeIndex slicing with boolean Index raises TypeError (#22852)

toobaz committed:
BUG: Return KeyError for invalid string key (#23540)

closes: #22803

simonjayhawkins committed:
TST: coverage for skipped tests in io/formats/test_to_html.py (#22888)

datapythonista committed:
DOC: Updating DataFrame.join docstring (#23471)

TomAugspurger committed:
ENH: Support EAs in Series.unstack (#23284)

gfyoung committed:
BUG: Fix error message for invalid HTML flavor (#23550)

The flavors were not rendering properly in the
string formatting.

Closes gh-23549.
Follow-up to gh-17660.

jbrockmendel committed:
remove uses of (ts)?lib.(NaT|iNaT|Timestamp) (#23562)

toobaz committed:
BUG: raise if invalid freq is passed (#23546)

jschendel committed:
TST: Fix dtype mismatch on 32bit in IntervalTree get_indexer test (#23468)

TomAugspurger committed:
Preserve EA dtype in DataFrame.stack (#23285)

TomAugspurger committed:
CI: Auto-cancel redundant builds (#23523)

mroeschke committed:
ENH: Support writing timestamps with timezones with to_sql (#22654)

Commit:
DOC: Added note about groupby excluding Decimal columns by default (#18953)

datapythonista committed:
Fixes to make validate_docstrings.py not generate warnings or unwanted output (#23552)

jorisvandenbossche committed:
Update description of Index._values/values/ndarray_values (#23507)

jbrockmendel committed:
TST: Tests and Helpers for Datetime/Period Arrays (#23502)

qwhelan committed:
PERF: define is_all_dates to shortcut inadvertent copy when slicing an IntervalIndex (#23591)

TomAugspurger committed:
PERF: Speeds up creation of Period, PeriodArray, with Offset freq (#23589)

FHaase committed:
DOC/CI: Add linting to rst files, and fix issues (#23381)

jorisvandenbossche committed:
CLN: remove values attribute from datetimelike EAs (#23603)

jschendel committed:
TST: Use intp as expected dtype in IntervalIndex indexing tests (#23609)

anjsudh committed:
ENH: Support for partition_cols in to_parquet (#23321)

* closes #23283

jmrr committed:
DOC: Updating Series.resample and DataFrame.resample docstrings (#23197)

gfyoung committed:
MAINT: tm.assert_raises_regex --> pytest.raises (#23592)

* MAINT: tm.assert_raises_regex --> pytest.raises

pytest.raises has all of the functionality that
we need from tm.assert_raises_regex.

Closes gh-16521.

* Don't remove, just deprecate assert_raises_regex

* CLN: Test cleanups and follow-ups

meiermark committed:
DOC: Remove incorrect periods at the end of parameter types (#23600)

datapythonista committed:
DOC: Fixes to docstring to add validation to CI (#23560)

TomAugspurger committed:
TST: Fix integer ops comparison test (#23619)

The `op(Series[integer], other)` path was being tested twice.
The `op(IntegerArray, other)` path was not being tested.

Closes https://github.com/pandas-dev/pandas/issues/22096

TomAugspurger committed:
TST: Unskip some Categorical Tests (#23613)

thoo committed:
DOC: Fix Order of parameters in docstrings (#23611)

jbrockmendel committed:
CLN: use float64_t consistently instead of double, double_t (#23583)

TomAugspurger committed:
API: DataFrame.__getitem__ returns Series for sparse column (#23561)

closes https://github.com/pandas-dev/pandas/issues/23559

gfyoung committed:
BUG: Delegate more of Excel parsing to CSV (#23544)

The idea is that we read the Excel file,
get the data, and then let the TextParser
handle the reading and parsing.

We shouldn't be doing a lot of work that
is already defined in parsers.py

In doing so, we identified several bugs:

* index_col=None was not being respected

* usecols behavior was inconsistent with
that of read_csv for list of strings and
callable inputs

* usecols was not being validated as proper
Excel column names when passed as a string.

Closes gh-18273.
Closes gh-20480.

jbrockmendel committed:
BUG: Casting tz-aware DatetimeIndex to object-dtype ndarray/Index (#23524)

jbrockmendel committed:
BUG/REF: TimedeltaIndex.__new__ (#23539)

datapythonista committed:
BUILD: Simplifying contributor dependencies (#23522)

jbrockmendel committed:
TST: Fix xfailing DataFrame arithmetic tests by transposing (#23620)

VincentLa committed:
DOC: Enhancing pivot / reshape docs (#21038)

kprestel committed:
CLN:Remove unused **kwargs from user facing methods (#23249)

datapythonista committed:
CI: Check in the CI that assert_raises_regex is not being used (#23627)

Follow-up to gh-23592.

jbrockmendel committed:
CLN: datetimelike arrays: isort, small reorg (#23587)

TomAugspurger committed:
BUG: astype fill_value for SparseArray.astype (#23547)

jtkiley committed:
More helpful Stata string length error. (#23629)

gfyoung committed:
DEPR: Deprecate usecols as int in read_excel (#23635)

Follow-up to gh-23544.

gfyoung committed:
BUG: Don't over-optimize memory with jagged CSV (#23527)

With jagged CSV's, we risk being too quick
to dump memory that we need to allocate
because previous chunks would have
indicated much larger rows than we can
anticipate in subsequent chunks.

Closes gh-23509.

datapythonista committed:
DOC: Adding validation of the section order in docstrings (#23607)

* Adding validation of the section order in docstrings

* Updating allowed sections

jbrockmendel committed:
CLN: remove incorrect usages of com.AbstractMethodError (#23625)

TomAugspurger committed:
DOC: avoid SparseArray.take error (#23637)

Closes https://github.com/pandas-dev/pandas/issues/22215

SparseArray.take not accepting scalars is already in 0.24.0.txt

jorisvandenbossche committed:
DOC: clean-up recent doc errors/warnings (#23636)

gueletk committed:
DOC: Fix name of the See Also section titles in docstrings (#23653)

benoxoft committed:
CI: Allow to compile docs with ipython 7.11 #22990 (#23655)

jschendel committed:
TST: IntervalTree.get_loc_interval should return platform int (#23660)

gfyoung committed:
CLN: Move to_excel to generic.py (#23656)

WillAyd committed:
Add to_flat_index method to MultiIndex (#22866)

gfyoung committed:
BUG: Fix read_excel w/parse_cols & empty dataset (#23661)

Closes gh-9208.

sechilds committed:
DOC: Update is_sparse docstring (#19983)

* BUG: Identify SparseDataFrame as sparse

The is_sparse function checks to see if
an array-like is spare by checking to see
if it is an instance of ABCSparseArray or
ABCSparseSeries. This commit adds
ABCSparseDataFrame to that list -- so it
can detect that a DataFrame (which is an
array-like object) is sparse.

Added a test for this.

* Revert "BUG: Identify SparseDataFrame as sparse"

This reverts commit 10dffd1452b50af101f26b331cbfabd3ea217434.

The previous commit's change was not necessary.
Will add a docstring to clarify the behaviour of the method.

* DOC: Revise is_sparce docstring

Clean up the docstring for is_sparse so it confirms to the
documentation style guide.

Add additional examples and clarify that is_sparse
expect a 1-dimensional array-like.

* DOC: Adjust is_sparse docstring.

Responding to pull request comments.

gfyoung committed:
DOC: Surface / doc mangle_dupe_cols in read_excel (#23678)

xref gh-10523.

justinchan23 committed:
Fix errorbar visualization (#23674)

myles committed:
DOC: Accessing files from a S3 bucket. (#23639)

gfyoung committed:
REF: Move Excel names parameter handling to CSV (#23690)

jschendel committed:
BUG: Fix Series/DataFrame.rank(pct=True) with more than 2**24 rows (#23688)

jreback commented:
lgtm. @TomAugspurger a glance if you can.

TomAugspurger commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -1312,6 +1312,7 @@ Notice how we now instead output ``np.nan`` itself instead of a stringified form
 - :func:`read_sas()` will correctly parse sas7bdat files with data page types having also bit 7 set (so page type is 128 + 256 = 384) (:issue:`16615`)
 - Bug in :meth:`detect_client_encoding` where potential ``IOError`` goes unhandled when importing in a mod_wsgi process due to restricted access to stdout. (:issue:`21552`)
 - Bug in :func:`to_string()` that broke column alignment when ``index=False`` and width of first column's values is greater than the width of first column's header (:issue:`16839`, :issue:`13032`)
+- Bug in :func:`to_string()` that caused representations of :class:`DataFrame` to not take up the whole window (:issue:`22984`)
```
This should be :class:\`DataFrame.to_string\` right? We don't have a top-level `pandas.to_string`.

jreback committed:
CI: raise clone depth limit on CI

TomAugspurger commented on pandas/tests/io/formats/test_format.py:
```diff
@@ -343,6 +343,16 @@ def test_repr_truncates_terminal_size(self):
 
         assert df2.columns[0] in result.split('\n')[0]
 
```
nitpick: we have a pytest fixture formock now, that does the try / except / skip done above.

I think it'd be cleanest to split this into a new test here, and accept the `mock` parameter.

```python
def test_repr_truncates_terminal_size_full(self, mock):
    ...
```

Any if you're feeling adventurous, you could change the try / except / skip mock import above to use the fixture as well. Not a big deal though.

jbrockmendel committed:
Implement _most_ of the EA interface for DTA/TDA (#23643)

jreback commented:
@JustinZhengBC need to merge master

JustinZhengBC committed:
Merge master

JustinZhengBC committed:
reapply changes

JustinZhengBC committed:
r

TomAugspurger commented:
Another hypothesis failure on azure. OK to merge @jreback? I can investigate the failing test separately.

jreback commented:
yep

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into JustinZhengBC-BUG-22984

TomAugspurger commented:
Merged master to fix the confict. May as well let the CI run.

JustinZhengBC commented:
@TomAugspurger green. Also thanks for fixing the merge issue

TomAugspurger commented:
I caused it my merging my own PR, so it's the least I could do :)
